build: Use XLIFF file to provide more context to Transifex translators

hebasto opened the pull request:
Currently, only a class name is provided to the Transifex translators as a context. Neither `disambiguation` parameter of the `tr()` function nor [translator comments](https://doc.qt.io/qt-5/i18n-source-translation.html#translator-comments), being included as XML elements to `*.ts` translation files, are not parsed by the Transifex due to its [limited support](https://docs.transifex.com/formats/qt-ts) of such files.

This PR makes possible to provide all of the context details via an intermediate [XLIFF](https://docs.transifex.com/formats/xliff) translation file.

With this PR `make -C src translate` produces the `src/qt/locale/bitcoin_en.xlf` file which must be provided to the Transifex as a translation source instead of `src/qt/locale/bitcoin_en.ts`.

Closes #21465.

An example translatable string with additional `<context>` and `<note>` XML elements: https://github.com/bitcoin/bitcoin/blob/35d52397e72f3ab96a7797148666b501d50b445d/src/qt/locale/bitcoin_en.xlf#L126-L132

hebasto committed:
build: Use XLIFF file to provide more context to Transifex translators

Details: https://docs.transifex.com/formats/xliff

hebasto committed:
build: Add Qt lconvert tool to depends

hebasto committed:
qt [experimental]: Add a translation comment and a disambiguation string

The goal is to see the way the Transifex presents the added items to
translators using an intermediate XLIFF translation file.

hebasto commented:
Hoping on "22.0" milestone assigned to this PR :)

laanwj commented on .gitignore:
```diff
@@ -56,6 +56,8 @@ src/qt/bitcoin-qt.creator
 src/qt/bitcoin-qt.files
 src/qt/bitcoin-qt.includes
+# An intermediate XLIFF translation file.
+src/qt/locale/bitcoin_es.xlf
```
s/es/en/ ?

hebasto replied:
Thanks! Fixed :facepalm:

hebasto commented:
CI is green now :)

hebasto commented:
#### Guix builds:
```
$ find guix-build-$(git rev-parse --short=12 HEAD)/output/ -type f -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum
4aacdb5fc0138a0587939be1f9c47bbd90e8d5f58d26ff632b3d3e0eab56a323  guix-build-347a45d40c0d/output/aarch64-linux-gnu/bitcoin-347a45d40c0d-aarch64-linux-gnu-debug.tar.gz
32ba44b565442555f61bbcadef67f9321ad62937be802459aed1e89ada0b54eb  guix-build-347a45d40c0d/output/aarch64-linux-gnu/bitcoin-347a45d40c0d-aarch64-linux-gnu.tar.gz
874df98fa3e5cfc1b9be6b1838c7bb35496652e7d000d65fa313a92fca459174  guix-build-347a45d40c0d/output/arm-linux-gnueabihf/bitcoin-347a45d40c0d-arm-linux-gnueabihf-debug.tar.gz
2958150120f71dfe7f8f342e5d5fda8adb50467f01e22b0587f7b16a91de125c  guix-build-347a45d40c0d/output/arm-linux-gnueabihf/bitcoin-347a45d40c0d-arm-linux-gnueabihf.tar.gz
87646c550c3b32df071464f608c86b468edb2b227321147fc1085c9dda17f518  guix-build-347a45d40c0d/output/dist-archive/bitcoin-347a45d40c0d.tar.gz
04da2a653141d61501b19596f96240abee2f89367e730cc3bcf201bb61975cff  guix-build-347a45d40c0d/output/powerpc64-linux-gnu/bitcoin-347a45d40c0d-powerpc64-linux-gnu-debug.tar.gz
c8948317ee27408c7e7817c075b0559419990f1ecc06d8c88829cfd67034c048  guix-build-347a45d40c0d/output/powerpc64-linux-gnu/bitcoin-347a45d40c0d-powerpc64-linux-gnu.tar.gz
f0e4a519299e72a7d2feee069a709d5dda4dbbb669ea84b131a1d2b479963d30  guix-build-347a45d40c0d/output/powerpc64le-linux-gnu/bitcoin-347a45d40c0d-powerpc64le-linux-gnu-debug.tar.gz
f9db97a01e7b153882b5ca8b89ae51862424414c7645f1f1413b1f9d6fa0fcad  guix-build-347a45d40c0d/output/powerpc64le-linux-gnu/bitcoin-347a45d40c0d-powerpc64le-linux-gnu.tar.gz
a7c99c2045c5a12277a49c1aa08a5fd8f55252f2db9661d364c1b41e0afd8c12  guix-build-347a45d40c0d/output/riscv64-linux-gnu/bitcoin-347a45d40c0d-riscv64-linux-gnu-debug.tar.gz
25d35e70ceeccd5ae3a1d1ab389b27192e6ebe217d5960f2e9a622787a5bf762  guix-build-347a45d40c0d/output/riscv64-linux-gnu/bitcoin-347a45d40c0d-riscv64-linux-gnu.tar.gz
5b2483d05f4fd393cf6ed5882001fc420dcf1add27200c171059dd2bb8ec2a27  guix-build-347a45d40c0d/output/x86_64-apple-darwin18/bitcoin-347a45d40c0d-osx-unsigned.dmg
534c838a6112ac87f9288aea585cf05f5a5450b58de74a4ad9388a14a4b56019  guix-build-347a45d40c0d/output/x86_64-apple-darwin18/bitcoin-347a45d40c0d-osx-unsigned.tar.gz
2506ffbc585f637d3b0587f9cee737e15df83bdaaca1d8f2373dcf21f55179cc  guix-build-347a45d40c0d/output/x86_64-apple-darwin18/bitcoin-347a45d40c0d-osx64.tar.gz
5d773c27ae724623a9a14538bddc23acbec9df49bef61306c943ba4c561fc70e  guix-build-347a45d40c0d/output/x86_64-linux-gnu/bitcoin-347a45d40c0d-x86_64-linux-gnu-debug.tar.gz
bc83d9f50b2d6d7c9dfc190cf62b22e1e6b152718edb2392e4f03b64e2fd01ac  guix-build-347a45d40c0d/output/x86_64-linux-gnu/bitcoin-347a45d40c0d-x86_64-linux-gnu.tar.gz
5048bc3cedb9454967213a7d9298672f0fc35ebe2097cbd21312d533542b8f9c  guix-build-347a45d40c0d/output/x86_64-w64-mingw32/bitcoin-347a45d40c0d-win-unsigned.tar.gz
414ce9df468cc0668e63f8ef6e2277f93481bb068423bb0ba073b713d455aa67  guix-build-347a45d40c0d/output/x86_64-w64-mingw32/bitcoin-347a45d40c0d-win64-debug.zip
bc364ddd8445d01fcc98759c8233a6243c0216505e68260d4316ce3d8be346e7  guix-build-347a45d40c0d/output/x86_64-w64-mingw32/bitcoin-347a45d40c0d-win64-setup-unsigned.exe
25b1f4390d2d937e38f1b44f6238202ddb6403b5e80c2474eef2f7eef94f83db  guix-build-347a45d40c0d/output/x86_64-w64-mingw32/bitcoin-347a45d40c0d-win64.zip
```
#### Gitian builds:
- Linux
```
Generating report
7f6b9251d2a7f0991c91eb96b21466b9798ab6f715871c485f239ea2abff186e  bitcoin-347a45d40c0d-aarch64-linux-gnu-debug.tar.gz
513a2b47811a0f896700fca152b139eba3c7da9c36993609d64ecbe3ecf4cba7  bitcoin-347a45d40c0d-aarch64-linux-gnu.tar.gz
14ff66645b7bce2675dbba932b67003a80891c14078820bdee2293ede8ce7ef0  bitcoin-347a45d40c0d-arm-linux-gnueabihf-debug.tar.gz
3fcd16876a712c291639d096f5e2b6dea908955754177d2a3680b1067ac1047f  bitcoin-347a45d40c0d-arm-linux-gnueabihf.tar.gz
398dab46c26041c77ff7e767bd0fb3f3f2f059b74995ce5f063b1297af20327d  bitcoin-347a45d40c0d-powerpc64-linux-gnu-debug.tar.gz
c041f14cfaf63ffd4fe753ad323eb1efd64e7fc4fc9ba7a2a208407380bb0d6e  bitcoin-347a45d40c0d-powerpc64-linux-gnu.tar.gz
dc270956eba3c52b481a6a8cae9fd3250d3b460ae648938e3f32ac53c532da0c  bitcoin-347a45d40c0d-powerpc64le-linux-gnu-debug.tar.gz
889142ea3376f5429996771ed1de5a9ac4eba5f595f976ab877841903439a751  bitcoin-347a45d40c0d-powerpc64le-linux-gnu.tar.gz
93cf2e3fd95064cced38db94dc388d50e1a5cd603e70a4a97bb71f3155282a1b  bitcoin-347a45d40c0d-riscv64-linux-gnu-debug.tar.gz
20413deb68c29eec0f8e07e117dde4a973986ee43025125b498518f718ad09a0  bitcoin-347a45d40c0d-riscv64-linux-gnu.tar.gz
3e01f68cfb4c12008b54eee253ce022df28d844d347a3f7382bb2509e09cd092  bitcoin-347a45d40c0d-x86_64-linux-gnu-debug.tar.gz
0ac8dc6d92a46d70de9a6f295075b1fa5c2ee6da7631238049e7a00d10fd6857  bitcoin-347a45d40c0d-x86_64-linux-gnu.tar.gz
87646c550c3b32df071464f608c86b468edb2b227321147fc1085c9dda17f518  src/bitcoin-347a45d40c0d.tar.gz
e4da41bd5eed8c447cd53a64f0c516d53e8aa7551a7df6c1c8ed3fef53c4f6ad  bitcoin-core-linux-22-res.yml
Done.
```
- Windows
```
Generating report
7c9ffd0c4afd4fc93a265ab77c8df6793c7e3a9d63d2ff106ce106131fed6784  bitcoin-347a45d40c0d-win-unsigned.tar.gz
298991c5a0b7be19a00d61e8382c90b710de3fc100f06f880c3c7781ecb1ad15  bitcoin-347a45d40c0d-win64-debug.zip
016aabe2d3d78a1e791f8b6d24c9592aa44b52b5a1934d5aaa8d19a81ebde1b5  bitcoin-347a45d40c0d-win64-setup-unsigned.exe
b29b0ca9e31bd0066eff36bb920d0699e62a97134af838002ffc7de5dc57199a  bitcoin-347a45d40c0d-win64.zip
87646c550c3b32df071464f608c86b468edb2b227321147fc1085c9dda17f518  src/bitcoin-347a45d40c0d.tar.gz
420519f953cf33ad2a6eb64f57284bc3bafc4c4d2d5898206e86890e969341a1  bitcoin-core-win-22-res.yml
Done.
```
- macOS
```
Generating report
f05f7feaef1ee495de27ba80d673ca63d91b812f617436c759102c46e9b42be9  bitcoin-347a45d40c0d-osx-unsigned.dmg
f3c966da3ab7516f736bc5471bcf20ee9530fb8d37b3ab05e86b4a93597e88cb  bitcoin-347a45d40c0d-osx-unsigned.tar.gz
ce3ad3701fe2fbc54f31e5b095f9fe6a1e038c3d84467c4a4d0e7252ef067c76  bitcoin-347a45d40c0d-osx64.tar.gz
87646c550c3b32df071464f608c86b468edb2b227321147fc1085c9dda17f518  src/bitcoin-347a45d40c0d.tar.gz
2809ec177a540dcd4531ad81ce45b3ed2b4d39ff37f901646f0a411a5dbd1de5  bitcoin-core-osx-22-res.yml
Done.
```

ysangkok commented:
I verified that `make -C src translate` produces an XML file with an extracomment with the error message in the comment. I verified that `build-unix.md` suggests the package that includes `lrelease`. And I verified that the `configure` script errors if `--with-gui` is specified and `qttools5-dev-tools` is not installed on `Ubuntu 20.04.2 LTS`

laanwj commented on .gitignore:
```diff
@@ -56,6 +56,8 @@ src/qt/bitcoin-qt.creator
 src/qt/bitcoin-qt.files
 src/qt/bitcoin-qt.includes
+# An intermediate XLIFF translation file.
+src/qt/locale/bitcoin_en.xlf
```
Transifex auto-updates from the branch. If you want transifex to pick this file up instead of the `.ts` it needs to be checked in, not added to `.gitignore`.

hebasto replied:
Thanks! Updated.

hebasto committed:
Add bitcoin_en.xlf intermediate translation file to the repo

As the bitcoin_en.xlf file is created by the `make -C src translate`
command, other translation-related files are also updated.

hebasto commented:
Updated 347a45d40c0dcad2a2286495e661cfffa7af329b -> 35d52397e72f3ab96a7797148666b501d50b445d ([pr21694.04](https://github.com/hebasto/bitcoin/commits/pr21694.04) -> [pr21694.05](https://github.com/hebasto/bitcoin/commits/pr21694.05), [diff](https://github.com/hebasto/bitcoin/compare/pr21694.04..pr21694.05)):

- addressed @laanwj's [comment](https://github.com/bitcoin/bitcoin/pull/21694#discussion_r616493850):
> Transifex auto-updates from the branch. If you want transifex to pick this file up instead of the `.ts` it needs to be checked in, not added to `.gitignore`.

hebasto commented:
The pr description updated with an example of a translatable string presentation in the `bitcoin_en.xlf` file.

laanwj commented:
ACK 35d52397e72f3ab96a7797148666b501d50b445d
Checked that the same `src/qt/locale/bitcoin_en.xlf` is generated locally.

hebasto commented:
The actual results are presented in https://github.com/bitcoin-core/gui/issues/287.
